Mathlib/Tactic/Linter/FlexibleLinter.lean

```diff
@@ -0,0 +1,380 @@
+  match stx with
+    | stx@(.node _ ``Lean.Parser.Tactic.location loc) =>
+      if all? stx then {} else (loc.map toStained).foldl (·.merge ·) {}
+    | .node _ _ args => (args.map (getStained · all?)).foldl (·.merge ·) {}
+    | _ => default
+
+/-- `getStained! stx` expects `stx` to be an argument of a node of `SyntaxNodeKind`
+`Lean.Parser.Tactic.location`.
+Typically, we apply `getStained!` to the output of `getLocs`.
+
+It returns the `HashSet` of `Stained` determined by the locations in `stx`.
+
+The only difference with `getStained stx`, is that `getStained!` never returns `{}`:
+if `getStained stx = {}`, then `getStained' stx = {.goal}`.
+
+This means that tactics that do not have an explicit "`at`" in their syntax will be treated as
+acting on the main goal. -/
+def getStained! (stx : Syntax) (all? : Syntax → Bool := fun _ ↦ false) : HashSet Stained :=
+  let out := getStained stx all?
+  if out.size == 0 then {.goal} else out
+
+/-- `Stained.toFMVarId mv lctx st` takes a metavariable `mv`, a local context `lctx` and
+a `Stained` `st` and returns the array of pairs `(FVarId, mv)`s that `lctx` assigns to `st`
+(the second component is always `mv`):
+* if `st` "is" a `Name`, returns the singleton of the `FVarId` with the name carried by `st`;
+* if `st` is `.goal`, returns the singleton `#[default]`;
+* if `st` is `.wildcard`, returns the array of all the `FVarId`s in `lctx` with also `default`
+  (to keep track of the `goal`).
+-/
+def Stained.toFMVarId (mv : MVarId) (lctx: LocalContext) : Stained → Array (FVarId × MVarId)
+  | name n   => match lctx.findFromUserName? n with
+                  | none => #[]
+                  | some decl => #[(decl.fvarId, mv)]
+  | goal     => #[(default, mv)]
+  | wildcard => (lctx.getFVarIds.push default).map (·, mv)
+
+/-- `SyntaxNodeKind`s that are mostly "formatting": mostly they are ignored
+because we do not want the linter to spend time on them.
+The nodes that they contain will be visited by the linter anyway. -/
+def combinatorLike : HashSet Name :=
+  { -- "continuators": these typically wrap other tactics inside them.
+    -- The linter ignores the wrapper, but does recurse into the enclosed tactics
+    ``Lean.Parser.Tactic.tacticSeq1Indented,
+    ``Lean.Parser.Tactic.tacticSeq,
+    ``Lean.Parser.Term.byTactic,
+    `by,
+    ``Lean.Parser.Tactic.tacticTry_,
+    `choice,  -- involved in `first`
+    ``Lean.Parser.Tactic.allGoals,
+    `Std.Tactic.«tacticOn_goal-_=>_»,
+    ``Lean.Parser.Tactic.«tactic_<;>_»,
+    ``cdotTk,
+    ``cdot,
+    -- "stopper tactics": their effect is analogous to "continuators", but they do not wrap
+    -- other tactics inside them.  Some are "finishing" tactics, so there really is nothing beyond
+    -- them in their branch of the InfoTree.
+    ``Lean.Parser.Tactic.tacticSorry,
+    ``Lean.Parser.Tactic.tacticRepeat_,
+    ``Lean.Parser.Tactic.tacticStop_,
+    `Mathlib.Tactic.Abel.abelNF,
+    `Mathlib.Tactic.RingNF.ringNF }
```

Comment:
Why are `ringNF` and `abelNF` stoppers (and not flexible tactics)?

Reply:
Looks like the intention is that they stop the spread of stain. So for example from the test, `simp; abel_nf; assumption` would be allowed because `abel_nf` puts everything in normal form. That's probably the best way to categorize them, but perhaps the comment above could be clearer about that.

Reply:
Indeed, Anne's interpretation is what I was hoping to achieve.  Part of the confusion may be that this started being a "flag non-terminal `simp`" linter, before becoming "start by staining anything that `simp` touches until it is touched by a
* rigid tactic that would likely break with "small" changes in its input (and then complain);
* a stopper that produces some normal form and should have a stable output (and then stop staining further downstream code)"

linter.

I'll try to make the names and descriptions clearer though!

Reply:
Ah I understand, thanks!